soc/samr21: Initial support

benpicco opened the pull request:
The samr21 is an ARM Cortex-M0 part very simmilar to the samd21, but comes with a on-chip radio (AT86RF233) connected internally via SPI.

This adds support for the SoC family and the SAM R21 Xplained Pro eveluation board.
Most of it is copy & paste from `atmel_sam0/samd21`.

The AT86RF233 radio is not supported yet at this point.

dbkinder commented on boards/arm/atsamr21_xpro/doc/index.rst:
```diff
@@ -0,0 +1,208 @@
+Overview
+********
+
+The `SAMR21 Xplained Pro` is a compact evaluation board by Atmel featuring a
```
```suggestion
The SAM R21 Xplained Pro is a compact evaluation board by Atmel featuring a
```

dbkinder commented on boards/arm/atsamr21_xpro/doc/index.rst:
```diff
@@ -0,0 +1,208 @@
+=====
+
+The SAMR21 SoC includes an on-chip AT86RF233 radio. It is internally
+connected via SPI and some GPIO pins and behaves in exact the same way as
```
```suggestion
connected via SPI and some GPIO pins and behaves the same way as
```

dbkinder commented on boards/arm/atsamr21_xpro/doc/index.rst:
```diff
@@ -0,0 +1,208 @@
+Flashing
+========
+
+#. Build the Zephyr kernel and the ``hello_world`` sample application:
```
This change will link to the hello world sample doc:
```suggestion
#. Build the Zephyr kernel and the :ref:`hello_world` sample application:
```

dbkinder reviewed (changes requested):
a few doc changes ...

mnkp commented:
This PR looks good overall, thanks @benpicco, but it needs to be split in three parts:
- ASF files provided by Atmel located in ext/
- samr21 soc files located in soc/
- atsamr21_xpro board support located in board/

The commit messages have to follow documentation guidelines [Identifying Contribution Origin](https://docs.zephyrproject.org/latest/contribute/index.html#identifying-contribution-origin). 

The header files provided by Atmel often have mixed line endings. We should change them to the Unix '\n'. Otherwise the first commit shouldn't contain any modifications to the original files.

dbkinder reviewed (approved):
+1 for doc changes, thanks!

benpicco committed:
ext: Import Atmel SAMR21 header files from ASF library

Atmel Software Framework (ASF) provides a set of low-level header
files that give access to different hardware peripherals of Atmel's
ICs.

Origin: Atmel SAMR21 Series Device Support (1.1.72)
License: Apache-2.0
URL: http://packs.download.atmel.com/Atmel.SAMR21_DFP.1.1.72.atpack
Purpose: Introduction of ASF for the SAM0 series.
Maintained-by: External

Signed-off-by: Benjamin Valentin <benpicco@googlemail.com>

benpicco committed:
soc: atmel: add SAMR21

Adds Atmel SAMR21 soc which is based on SAMD21, but with a AT86RF233
radio connected internally via SPI.

The AT86RF233 is not yet supprted by Zephyr at this point.

This code is very much copy & paste from atmel_sam0/samd21

Signed-off-by: Benjamin Valentin <benpicco@googlemail.com>

benpicco committed:
boards: atsamr21_xpro: Add SAM R21 Xplained Pro Evaluation Kit

Add the Atmel SAM R21 Xplained Pro Evaluation Kit to zephyr.

So far, UART, SPI, I2C (depends on #14128), debug LED and user button
have been tested.

Signed-off-by: Benjamin Valentin <benpicco@googlemail.com>

mnkp commented on boards/arm/atsamr21_xpro/atsamr21_xpro_defconfig:
```diff
@@ -0,0 +1,23 @@
+CONFIG_SOC_ATMEL_SAMD_XOSC32K=y
+CONFIG_SOC_ATMEL_SAMD_XOSC32K_AS_MAIN=y
+# CONFIG_SPI is not set
+# CONFIG_SPI_SAM0 is not set
```
Unused code should be removed. However, board documentation claims that SPI is supported.

benpicco replied:
I could create a samr21.dtsi with
```
#include <atmel/samd21.dtsi>
&sercom4 {
       status = "ok";
       compatible = "atmel,sam0-spi";
       dipo = <0>;
       dopo = <2>;
};
```
which is the internal SPI where the radio is attached to like in [nxp_kw2xd.dtsi](https://github.com/zephyrproject-rtos/zephyr/blob/master/dts/arm/nxp/nxp_kw2xd.dtsi#L249). There a at86rf233@0 node could be attached later.

I could also just map one of the remaining SerComs to SPI on the debug header of the board.

mnkp replied:
If the code is not used it should be removed. Also - the rules are a bit fuzzy here - I believe that only basic functionality should be enabled in this file. @galak will know the definite answer.

benpicco replied:
Ok, I removed it for now until, can add it back later when support for the radio is in place.

mnkp commented on boards/arm/atsamr21_xpro/doc/index.rst:
```diff
@@ -0,0 +1,208 @@
+| USART     | on-chip    | Serial ports                         |
++-----------+------------+--------------------------------------+
+| SPI       | on-chip    | Serial Peripheral Interface ports    |
++-----------+------------+--------------------------------------+
```
Commit message states that "only UART, debug LED and user button are supported". Here we have also WDT, SPI.

benpicco replied:
s/supported/tested/

mnkp reviewed (changes requested):
There are several minor issues. Overall it looks good.

One thing which needs to be globally fixed are copyright headers. You'll need to include your name there. I'm only not sure what should happen with the name of the original author. @nashif or @galak should know.

mnkp commented on soc/arm/atmel_sam0/samr21/soc.h:
```diff
@@ -0,0 +1,57 @@
+ */
+
+#ifndef ZEPHYR_ATMEL_SAMD_SOC_H_
+#define ZEPHYR_ATMEL_SAMD_SOC_H_
```
`ZEPHYR_ATMEL_SAMR21_SOC_H_`

benpicco replied:
OK

mnkp commented on boards/arm/atsamr21_xpro/atsamr21_xpro.dts:
```diff
@@ -0,0 +1,58 @@
+ */
+
+/dts-v1/;
+#include <atmel/samd21.dtsi>
```
We should include `atmel/samr21.dtsi`. The "soc: atmel: add SAMR21" commit should add the required .dtsi file.

benpicco replied:
Ok.

mnkp commented on soc/arm/atmel_sam0/common/soc_power.c:
```diff
@@ -0,0 +1,86 @@
+	__WFI();
+}
+
+static inline void _sleep(void) {
```
Let's call this function `deep_sleep`. Sleep normally means power mode where we don't loose execution context.

benpicco replied:
OK, I named it `sam0_deep_sleep()`

mnkp commented on soc/arm/atmel_sam0/common/soc_power.c:
```diff
@@ -0,0 +1,86 @@
+#include <logging/log.h>
+LOG_MODULE_DECLARE(soc, CONFIG_SOC_LOG_LEVEL);
+
+static inline void _idle(uint8_t mode) {
```
I'm a bit confused myself about usage rules of leading underscore in function names, however we generally don't use them for static functions.

benpicco replied:
I renamed the function to `sam0_idle()`

mnkp reviewed (commented):
Power management system has been recently updated. As you probably have noticed by now this PR needs to be rebased on top of the latest master.

mnkp commented on soc/arm/atmel_sam0/common/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,35 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+if SOC_FAMILY_SAM0
```
That's a good idea but we probably don't need a separate `Kconfig.defconfig.series` file. I would put this section directly into `soc/arm/atmel_sam0/common/Kconfig.defconfig`.

benpicco replied:
Ok, I was just copying what `st_stm32` was doing.

mnkp commented on soc/arm/atmel_sam0/common/power.h:
```diff
@@ -0,0 +1,35 @@
+/*
```
This file is no longer necessary and can be removed. Functions declarations are available via `include/power.h`.

benpicco replied:
removed it.

benpicco replied:
oops, sorry - I hope I didn't miss anything else, but I think here I just used `rm` instead of `git rm` and forgot to commit it.

mnkp reviewed (commented):
LGTM with minor comments.

Unfortunately, there is an ongoing work regarding Power Management API (#13632) which will break current implemenation. When #13632 is merged one will need to run a few string replacement operations in `soc/arm/atmel_sam0/common/power.c`.

benpicco commented:
I replaced the strings with their new versions from #13632

`dts_fixup.h` is copied verbatim from the samd21 folder, it's unclear who the original author was as the first git commit to it was a move from another directory.

Are Codeowners needed for files in `ext/hal`?

benpicco committed:
CODEOWNERS: Add code ownership for Atmel samr21 code

Take ownership of external SDK files.

Signed-off-by: Benjamin Valentin <benjamin.valentin@ml-pa.com>

benpicco commented:
@mnkp I replaced the strings with their new versions from #13632

The rest of the changes should be addressed now too.

benpicco commented on soc/arm/atmel_sam0/samr21/Kconfig.series:
```diff
@@ -0,0 +1,15 @@
+# Kconfig - Atmel SAMD21 MCU series
```
Ok.

benpicco commented on soc/arm/atmel_sam0/samr21/Kconfig.soc:
```diff
@@ -0,0 +1,31 @@
+# Kconfig - Atmel SAMR MCU series
```
Ok.

benpicco commented on soc/arm/atmel_sam0/samr21/Kconfig.soc:
```diff
@@ -0,0 +1,31 @@
+# SPDX-License-Identifier: Apache-2.0
+
+choice
+prompt "Atmel SAME21 MCU Selection"
```
Ok.

benpicco commented on soc/arm/atmel_sam0/samr21/CMakeLists.txt:
```diff
@@ -0,0 +1,6 @@
+# Makefile - Atmel SAMD21 MCU series.
```
Ok.

benpicco commented on soc/arm/atmel_sam0/samr21/Kconfig.series:
```diff
@@ -0,0 +1,15 @@
+	bool "Atmel SAMR21 MCU"
+	select CPU_CORTEX_M0PLUS
+	select SOC_FAMILY_SAM0
+	select SYS_POWER_LOW_POWER_STATES_SUPPORTED
```
Removed.

benpicco commented on soc/arm/atmel_sam0/samr21/soc.c:
```diff
@@ -0,0 +1,197 @@
+
+/**
+ * @file
+ * @brief Atmel SAMD MCU series initialization code
```
Ok.

benpicco commented on soc/arm/atmel_sam0/samr21/soc.h:
```diff
@@ -0,0 +1,57 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef _ATMEL_SAMD_SOC_H_
```
changed it to `ZEPHYR_ATMEL_SAMR21_SOC_H_`

benpicco commented:
Once #14005 #13717 and #14128 are merged they can be enabled here as well.

mnkp reviewed (commented):
@benpicco Sorry for a delay, I was busy. LGTM but some things have not yet been fixed. Maybe you didn't push all the local changes you did?

Also a minor thing, the "soc: atmel: auto-select peripherals on SAM0" could be squashed with "soc: atmel: add SAMR21" as it logically belongs there.

benpicco commented:
Sorry for resolving that prematurely - I'll check tomorrow if there are still any changes on my computer at work.

benpicco commented:
@mnkp  Ok, I found no more changes - squashed the commit.

jfischer-phytec-iot commented:
@benpicco would you like to add usb support for SAM R21 Xplained board? It should be the same as for SAMD21, also pinmuxing for DP and DM.

benpicco commented:
@jfischer-phytec-iot I added it to `pinmux.c` - that seems to be the only occurrence on atsamd21_xpro anyway.

benpicco commented:
I sqashed the fixups.

benpicco committed:
soc: sam0: enable RTC for samd21, samr21

The Atmel SAMD21 (and therefore also the SAMR21) comes with the same
RTC peripheral as the Atmel SAMD20.

Enable it in dts_fixup.h and enable it in the dts for samr21_xpro.

Signed-off-by: Benjamin Valentin <benpicco@googlemail.com>

benpicco commented:
To make rebasing easier, I dropped the unrelated commits into separate pull requests.

benpicco committed:
soc: sam0: Make init code common between all samd2x SoCs

The init routines are the same for SAMD20, SAMD21 and SAMR21, so
move them into common/ to not have three copies of the same code.

Signed-off-by: Benjamin Valentin <benjamin.valentin@ml-pa.com>

mnkp commented on soc/arm/atmel_sam0/common/soc_samd2x.c:
```diff
@@ -55,8 +55,13 @@ static void xosc32k_init(void)
 	u32_t fuse = *(u32_t *)FUSES_OSC32K_CAL_ADDR;
 	u32_t calib = (fuse & FUSES_OSC32K_CAL_Msk) >> FUSES_OSC32K_CAL_Pos;
+#else
+	u32_t fuse = *(u32_t *)FUSES_OSC32KCAL_ADDR;
```
This looks like a bug in the header files. Maybe we should report it to Microchip?

benpicco commented on soc/arm/atmel_sam0/samd20/CMakeLists.txt:
```diff
@@ -3,4 +3,4 @@
 # SPDX-License-Identifier: Apache-2.0
 
-zephyr_sources(soc.c)
+zephyr_sources(../common/soc_samd2x.c)
```
I've changed it so that `soc_samd2x.c` now gets selected based on whether ´CONFIG_SOC_SERIES_SAMD20`, `CONFIG_SOC_SERIES_SAMD21` or `CONFIG_SOC_SERIES_SAMR21` is set.

samd20/, samd21/ and samr21/ only contain header files and KConfig now.

jfischer-phytec-iot commented on boards/arm/atsamr21_xpro/atsamr21_xpro.yaml:
```diff
@@ -0,0 +1,12 @@
+toolchain:
+  - zephyr
+  - gnuarmemb
+  - xtools
```
(at least gpio and usb, i2c?)

```
supported:                                                                                                                                                                   
  - gpio                                                                                      
  - usb_device                                                                                

```

benpicco replied:
I'll add gpio & usb - i2c depends on #14128

benpicco replied:
Ah turns out this makes CI build some more projects with the board!
The build failure should be fixed by #14832 then. I'd rather have this merged first than manually enabling the driver implementations to the board default config.

benpicco replied:
It still fails!
#15584 should fix this :crossed_fingers:

mnkp commented on soc/arm/atmel_sam0/common/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+
+zephyr_sources_ifdef(CONFIG_SOC_SERIES_SAMD20 soc_samd2x.c)
+zephyr_sources_ifdef(CONFIG_SOC_SERIES_SAMD21 soc_samd2x.c)
+zephyr_sources_ifdef(CONFIG_SOC_SERIES_SAMR21 soc_samd2x.c)
```
Maybe we should write it as
```
zephyr_sources_ifdef(
  (CONFIG_SOC_SERIES_SAMD20 OR CONFIG_SOC_SERIES_SAMD21 OR 
   CONFIG_SOC_SERIES_SAMR21)
  soc_samd2x.c)
```

benpicco replied:
If I write it like that, `soc_samd2x.c` does not get build.
